### PR TITLE
Feature/reconnection v2 attempt 1

### PR DESCRIPTION
**Pull request by ipavlidakis**
- Each commit will be a separate PR targeting feature/reconnect-v2
- Joining calls 🟢
- Fast reconnect 🟢
- Rejoin 🟢
- Migration 🟢
- Call ended event handling 🟢
- Local Screensharing 🟢

Missing/Bugs:
- Visibillity modifier

### CODE REVIEW

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/ScreenShareMediaAdapter.swift [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            let localScreenShareMediaManager = localMediaManager as? LocalScreenShareMediaAdapter
+        else {
+            return
+        }
+
+        try await localScreenShareMediaManager.beginScreenSharing(
+            of: type,
+            ownCapabilities: ownCapabilities
+        ) { [weak self] in
+//            self?.queue.sync { [weak self] in
```
Comment (martinmitrevski) [inline Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/ScreenShareMediaAdapter.swift:122]: no op here?

File: Sources/StreamVideo/WebRTC/v2/StateMachine/Stages/WebRTCCoordinator+Joining.swift [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+            var participants = joinResponse
+                .callState
+                .participants
+                .map { $0.toCallParticipant() }
+                .reduce(into: [String: CallParticipant]()) { partialResult, participant in
+                    partialResult[participant.sessionId] = participant
+                }
+
+            // - WORKAROUND - TO BE REMOVED - BEGIN
+            if mergeWithLocalParticipants {
```
Comment (martinmitrevski) [inline Sources/StreamVideo/WebRTC/v2/StateMachine/Stages/WebRTCCoordinator+Joining.swift:326]: can you explain this in more details?

File: Sources/StreamVideo/WebRTC/v2/StateMachine/WebRTCCoordinator+StateMachine.swift [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        /// Waits for the next stage of the specified type after optionally skipping initial stages.
+        ///
+        /// - Parameters:
+        ///   - stageType: The expected type of the next stage.
+        ///   - dropFirst: The number of initial stages to skip. Defaults to 0.
+        /// - Returns: The next stage of the specified type.
+        /// - Throws: An error if the next stage is not of the expected type or if the state machine transitions to an error stage.
+        func nextStageShouldBe<S: Stage>(
```
Comment (martinmitrevski) [inline Sources/StreamVideo/WebRTC/v2/StateMachine/WebRTCCoordinator+StateMachine.swift:44]: We should remove the other duplicate methods above

File: Sources/StreamVideoSwiftUI/CallView/VideoParticipantsView.swift [hunk 4]
```diff
@@ -361,7 +361,7 @@ public struct VideoCallParticipantView: View {
                     }
                 }
             )
-            .id(participant.id)
+//            .id(participant.renderingId)
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/CallView/VideoParticipantsView.swift:364]: why do we remove it?

File: Sources/StreamVideoSwiftUI/CallView/VisibilityThresholdModifier.swift [hunk 5]
```diff
@@ -100,12 +100,13 @@ extension View {
...
         changeHandler: @escaping (Bool) -> Void
     ) -> some View {
-        modifier(
-            VisibilityThresholdModifier(
-                in: bounds,
-                threshold: threshold,
-                changeHandler: changeHandler
-            )
-        )
+//        modifier(
```
Comment (martinmitrevski) [inline Sources/StreamVideoSwiftUI/CallView/VisibilityThresholdModifier.swift:103]: would prefer if we keep this one in, instead of using onAppear/Disappear.

File: Sources/StreamVideo/Call.swift [hunk 6]
```diff
@@ -928,8 +928,8 @@ public class Call: @unchecked Sendable, WSEventsSubscriber {
     /// will be thrown to indicate that the operation is not supported.
     ///
     /// - Warning: A maximum of one output of each type may be added.
-    public func addCapturePhotoOutput(_ capturePhotoOutput: AVCapturePhotoOutput) throws {
-        try callController.addCapturePhotoOutput(capturePhotoOutput)
+    public func addCapturePhotoOutput(_ capturePhotoOutput: AVCapturePhotoOutput) async throws {
+        try await callController.addCapturePhotoOutput(capturePhotoOutput)
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/Call.swift:932]: Even though we could simply wrap the call internally to a Task, and that will help us by not needing to add the async. At the same time it would invalidate our API as we aren't able then to throw any error that occurred (as that requires waiting for the task to execute).

So that's as small as possible while remaining aligned with the needs of the method.

File: Sources/StreamVideo/Controllers/CallController.swift [hunk 7]
```diff
@@ -75,137 +197,131 @@ class CallController: @unchecked Sendable {
...
-            callSettings: settings,
-            ring: ring,
-            migratingFrom: migratingFrom
-        )
-        
-        setupStatsTimer()
-        
+        try await webRTCCoordinator.connect(callSettings: callSettings)
+//        try await webRTCClient.connect(callSettings: callSettings)
+        guard let response = try await joinCallResponseSubject.nextValue(dropFirst: 1, timeout: 15) else {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/Controllers/CallController.swift]: Because the first value we receive is the currentValue of the subject (on the first access the first value will be nil). We rather want to the next available value that will be received as the connect flow (we started on the live above) proceeds.

File: Sources/StreamVideo/Models/CallParticipant.swift [hunk 8]
```diff
@@ -107,21 +107,47 @@ public struct CallParticipant: Identifiable, Sendable, Equatable {
         self.audioLevels = audioLevels
         self.pin = pin
     }
-    
+
+    public static func == (lhs: CallParticipant, rhs: CallParticipant) -> Bool {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/Models/CallParticipant.swift:111]: So we can provide equality checks based also on the RTCVideoTrack equality as by default the RTCVideoTrack is relying on objc equality (so just memory address). While now we also check the trackId.

File: Sources/StreamVideo/WebRTC/PeerConnection.swift [hunk 9]
```diff
@@ -37,32 +40,114 @@ class PeerConnection: NSObject, RTCPeerConnectionDelegate, @unchecked Sendable {
...
+            sessionID: sessionId,
+            peerType: type,
+            peerConnection: pc,
+            sfuAdapter: sfuAdapter
+        )
         super.init()
-        self.pc.delegate = self
+//        self.pc.delegate = self
+
+        pc
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/PeerConnection.swift]: That file has now been removed.

File: Sources/StreamVideo/WebRTC/PeerConnection.swift [hunk 10]
```diff
@@ -6,25 +6,28 @@ import Foundation
...
-    case subscriber
-    case publisher
-}
+//enum PeerConnectionType: String {
+//    case subscriber
+//    case publisher
+//}
 
-class PeerConnection: NSObject, RTCPeerConnectionDelegate, @unchecked Sendable {
+class PeerConnection: NSObject, @unchecked Sendable {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/PeerConnection.swift]: Yeap! And we are finally done with it 🥳

File: Sources/StreamVideo/WebRTC/SfuMiddleware.swift [hunk 11]
```diff
@@ -11,93 +11,137 @@ class SfuMiddleware: EventMiddleware {
...
+
                 case let .callEnded(event):
                     log.debug("Received call ended event with reason \(event.reason)")
+
                 case let .participantUpdated(event):
                     await handleParticipantUpdated(event)
+
                 case .participantMigrationComplete:
                     log.debug("Participant migration complete")
+                default:
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/SfuMiddleware.swift]: This file is also been removed.

File: Sources/StreamVideo/WebRTC/SfuMiddleware.swift [hunk 12]
```diff
@@ -11,93 +11,137 @@ class SfuMiddleware: EventMiddleware {
...
             do {
-                guard case let .sfuEvent(event) = event else {
+                guard let self, case let .sfuEvent(event) = event else {
                     return
                 }
+                log.debug("Received an event \(event)", subsystems: .sfu)
                 switch event {
-                case let .subscriberOffer(event):
-                    await handleSubscriberEvent(event)
+//                case let .subscriberOffer(event):
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/SfuMiddleware.swift]: We do handle it [here](https://github.com/GetStream/stream-video-swift/pull/480/files#diff-2fec2097e6782a417260f8385709f6a4aa5f4aeaa6437f9159c898c4077fdc1dR528)

File: Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Connecting.swift [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                        .$connectionState
+                        .filter {
+                            switch $0 {
+                            case .authenticating:
+                                return true
+                            default:
+                                return false
+                            }
+                        }
+                        .nextValue(timeout: 5)
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Connecting.swift]: Yes that's the timeout we set. I have aligned those values with React but happy to revisit them (and organise them slightly better as now they are just magic numbers)

File: Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Joining.swift [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+                            }
+                        }
+                        .compactMap { (event: Stream_Video_Sfu_Event_SfuEvent.OneOf_EventPayload) in
+                            if case let .joinResponse(response) = event {
+                                return response
+                            } else {
+                                return nil
+                            }
+                        }
+                        .nextValue(timeout: 5)
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Joining.swift]: 100%

[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Joining.swift]: I'm going to do that in the subsequent PRs

File: Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Migrated.swift [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+                            switch $0 {
+                            case .authenticating:
+                                return true
+                            default:
+                                return false
+                            }
+                        }
+                        .nextValue(timeout: 10)
+
+                    try transition?(
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Migrated.swift]: Yeah i'm using 80 characters page width but in some cases i have used this kind of formatting to allow easier changes. We will update those in the subsequent PRs

File: Sources/StreamVideo/WebRTC/VideoCapturer.swift [hunk 16]
```diff
@@ -55,7 +55,7 @@ class VideoCapturer: CameraVideoCapturing {
     func startCapture(device: AVCaptureDevice?) async throws {
         try await withCheckedThrowingContinuation { continuation in
             guard let videoCapturer = videoCapturer as? RTCCameraVideoCapturer, let device else {
-                continuation.resume(throwing: ClientError.Unexpected())
+                continuation.resume()
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/VideoCapturer.swift:58]: The error here doesn't make much sense as the guard can only fail if the device is nil (which should trigger an error on the CameraCapturer). That being said, having an error is actually not problematic so I'm going to bring it back. Good catch!

File: Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Migrated.swift [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+                    guard
+                        let client = context.client,
+                        let migratingSFUAdapter = client.migratingSFUAdapter
+                    else {
+                        throw ClientError(
+                            "WebRCTAdapter instance not available."
+                        )
+                    }
+
+                    let response = try await client
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/StateMachine/Stages/WebRTCClient+Migrated.swift]: We will 100% do!

File: Sources/StreamVideo/WebRTC/WebRTCClient.swift [hunk 18]
```diff
@@ -646,248 +757,111 @@ class WebRTCClient: NSObject, @unchecked Sendable {
...
             }
         }
     }
-    
+
+    func completeMigration() async {
+        await sfuAdapter.disconnect()
+        sfuAdapter = nil
+        sfuAdapter = migratingSFUAdapter
+        sfuMiddleware.sfuAdapter = migratingSFUAdapter!
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/WebRTCClient.swift]: That was only during the migration between the WebRTCClient and WebRTCCoordinator. Right now there shouldn't be any places where we force unwrap

File: Sources/StreamVideo/WebRTC/WebRTCClient.swift [hunk 19]
```diff
@@ -1110,112 +1174,97 @@ class WebRTCClient: NSObject, @unchecked Sendable {
...
-                    log.error(error)
-                }
-            }
-        }
-        
-        return webSocketClient
-    }
-    
+
+    private var reconnectAttempt: UInt32 = 0
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/WebRTCClient.swift]: The file has been removed.

File: Sources/StreamVideo/WebRTC/v2/Extensions/Foundation/Publisher+TaskSink.swift [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public func receive(subscription: Subscription) {
+        subscription.request(.unlimited)
+    }
+
+    public func receive(_ input: Input) -> Subscribers.Demand {
+        if let disposableBag {
+            Task {
+                do {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/Extensions/Foundation/Publisher+TaskSink.swift]: I had to change the implementation here as it was causing memory leaks. The current one doesn't have any duplication.

File: Sources/StreamVideo/WebRTC/v2/Extensions/Foundation/Publisher+TaskSink.swift [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+                    try await task(input)
+                } catch {
+                    completion?(.failure(error as! Failure))
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/Extensions/Foundation/Publisher+TaskSink.swift]: Note: The implementation here has changed.

The reason why we did that before was that the Subscriber we had set needed a specific failure type which the compiler couldn't infer from the error in the do catch.
The new implementation doesn't use force unwrapping.

File: Sources/StreamVideo/WebRTC/v2/Extensions/Foundation/Publisher+AsyncStream.swift [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+import Foundation
+import Combine
+
+extension Publisher {
+
+    func asyncStreamSink(
+        receiveCompletion: @escaping (Subscribers.Completion<Self.Failure>) -> Void = { _ in },
+        _ handler: @escaping (Self.Output) -> Void
+    ) -> AnyCancellable {
+        if #available(iOS 15.0, *) {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/Extensions/Foundation/Publisher+AsyncStream.swift:14]: The `self.values` property on the Combine.Publisher has an API availability of iOS15+

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/Extensions/RTCPeerConnection+DelegatePublisher.swift [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+import StreamWebRTC
+
+extension RTCPeerConnection {
+    private enum AssociatedKeys {
+        static var delegatePublisher = "DelegatePublisher"
+        static var dispatchQueue = "DispatchQueue"
+    }
+
+    var subject: PassthroughSubject<RTCPeerConnectionEvent, Never> {
+        withUnsafePointer(to: &AssociatedKeys.delegatePublisher) { key in
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/Extensions/RTCPeerConnection+DelegatePublisher.swift:16]: Noted!

[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/Extensions/RTCPeerConnection+DelegatePublisher.swift:16]: (Not resolving so we have a reference)

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/Extensions/SignalServer+Subject.swift [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+        """
+        Type: \(type(of: self))
+            trackType: \(trackType)
+            muted: \(muted)
+        """
+    }
+}
+
+extension Stream_Video_Sfu_Signal_UpdateMuteStatesResponse: SignalServerEvent {
+    var description: String {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/Extensions/SignalServer+Subject.swift:130]: The protocol extension won't have access on each event's properties though (e.g some have an error some others don't)

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/LocalMediaAdapters/LocalNoOpMediaAdapter.swift [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+//  NoOpLocalMediaManager.swift
+//  StreamVideo
+//
+//  Created by Ilias Pavlidakis on 6/8/24.
+//
+
+import Foundation
+import Combine
+
+final class LocalNoOpMediaAdapter: LocalMediaAdapting {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/LocalMediaAdapters/LocalNoOpMediaAdapter.swift:8]: In that way we can use the same object (RTCPeerConnectionCoordinator) for Publisher and Subscriber while not having to use optional values.

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/MediaAdapter.swift [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+            with: settings,
+            ownCapabilities: ownCapabilities
+        )
+
+        try await videoMediaAdapter.setUp(
+            with: settings,
+            ownCapabilities: ownCapabilities
+        )
+
+        try await screenShareMediaAdapter.setUp(
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/MediaAdapter.swift:108]: Yeah, the mediaAdapter will throw an error that will be propagated up to the WebRTCCoordinatorStage and from there a disconnection flow will be triggered (where we will try to reconnect if possible).

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/MediaAdapter.swift [hunk 27]
```diff
@@ -0,0 +1,243 @@
...
+
+            group.addTask {
+                try await videoMediaAdapter.didUpdateCallSettings(settings)
+            }
+
+            group.addTask {
+                try await screenShareMediaAdapter.didUpdateCallSettings(settings)
+            }
+
+            while try await group.next() != nil {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/MediaAdapter.swift]: This is the way to wait for all tasks in the group to complete. This is different than doing 
```swift
try await audioMediaAdapter.didUpdateCallSettings(settings)
try await videoMediaAdapter.didUpdateCallSettings(settings)
try await screenSharingMediaAdapter.didUpdateCallSettings(settings)
```
as here things will run sequentially while in our group things run in parallel.

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/ScreenShareMediaAdapter.swift (hunk 1 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/MediaAdapters/ScreenShareMediaAdapter.swift:119]: That's not needed anymore and thus has been removed (local streams are managed by the transceiver) and remote tracks lifecycle is managed by the SFU events.

File: Sources/StreamVideo/WebRTC/v2/StateMachine/Stages/WebRTCCoordinator+Joining.swift (hunk 2 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/StateMachine/Stages/WebRTCCoordinator+Joining.swift]: That was a workaround needed because the SFU wasn't returning us the latest participant state whenever we migrated. That has now been fixed and thus the workaround has been removed.

File: Sources/StreamVideoSwiftUI/CallView/VideoParticipantsView.swift (hunk 4 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/CallView/VideoParticipantsView.swift:364]: That was to ensure that tracks aren't disappearing because views where stuck in a previous superview. I have reverted it.

File: Sources/StreamVideoSwiftUI/CallView/VisibilityThresholdModifier.swift (hunk 5 above)
Comment (ipavlidakis) [reply Sources/StreamVideoSwiftUI/CallView/VisibilityThresholdModifier.swift:103]: We will for sure keep it. I have spent too much time tweaking it. Leaving this comment unresolved so we have a reference and be able to restore in a subsequent PR

File: .github/workflows/cron-checks.yml [hunk 28]
```diff
@@ -2,8 +2,9 @@ name: Cron Checks
 
 on:
   schedule:
-    # Runs "At 01:00 every night"
-    - cron: '0 1 * * *'
+    # Runs "At 01:00 every night except weekends"
+    - cron: '0 1 * * 1-5'
```
Comment (martinmitrevski) [inline .github/workflows/cron-checks.yml:6]: why are these changes included?

File: CHANGELOG.md [hunk 29]
```diff
@@ -4,7 +4,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 # Upcoming
 
-### 🔄 Changed
+### ✅ Added
+- Participants (regular and anonymous) count, can be accessed - before or after joining a call - from the `Call.state.participantCount` & `Call.state.anonymousParticipantCount` respectively. [#496](https://github.com/GetStream/stream-video-swift/pull/496)
```
Comment (martinmitrevski) [inline CHANGELOG.md:8]: Actually, you should update the base branch with develop

File: DemoApp/Sources/Views/StatsView/DemoStatsView.swift [hunk 30]
```diff
@@ -278,7 +279,7 @@ private struct DemoStatView<Value: Comparable>: View {
                 .$statsReport
                 .receive(on: DispatchQueue.global(qos: .utility))
                 .map(valueTransformer)
-                .removeDuplicates()
+//                .removeDuplicates()
```
Comment (martinmitrevski) [inline DemoApp/Sources/Views/StatsView/DemoStatsView.swift:282]: Leftover

File: Sources/StreamVideo/Call.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (martinmitrevski) [reply Sources/StreamVideo/Call.swift:932]: ok, let's mention this in the changelog - best in this PR so we don't forget.

File: Sources/StreamVideo/CallKit/CallKitAdapter.swift [hunk 31]
```diff
@@ -20,6 +20,12 @@ open class CallKitAdapter {
         set { callKitService.iconTemplateImageData = newValue }
     }
 
+    /// The callSettings to use when joining a call (after accepting it on CallKit)
+    /// default: nil
```
Comment (martinmitrevski) [inline Sources/StreamVideo/CallKit/CallKitAdapter.swift:24]: what's up with this  character?

File: Sources/StreamVideo/CallState.swift [hunk 32]
```diff
@@ -227,6 +226,18 @@ public class CallState: ObservableObject {
             transcribing = true
         case .typeCallTranscriptionStoppedEvent:
             transcribing = false
+        case .typeCallMissedEvent:
```
Comment (martinmitrevski) [inline Sources/StreamVideo/CallState.swift:229]: this also shouldn't be here

File: Sources/StreamVideo/Controllers/CallController.swift [hunk 33]
```diff
@@ -75,137 +122,119 @@ class CallController: @unchecked Sendable {
...
+        try await webRTCCoordinator.connect(
+            callSettings: callSettings,
+            ring: ring
         )
-        
-        setupStatsTimer()
-        
+        guard
+            let response = try await joinCallResponseSubject
+            .nextValue(dropFirst: 1, timeout: 15)
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Controllers/CallController.swift:131]: if we have several cases like this, it might make sense to extract the timeout in a constant.

File: Sources/StreamVideo/WebRTC/SFUEventAdapter.swift [hunk 34]
```diff
@@ -0,0 +1,379 @@
...
+    ) {
+        self.sfuAdapter = sfuAdapter
+        self.stateAdapter = stateAdapter
+        observeEvents()
+    }
+
+    private func observeEvents() {
+        disposableBag.removeAll()
+
+        sfuAdapter
```
Comment (martinmitrevski) [inline Sources/StreamVideo/WebRTC/SFUEventAdapter.swift:27]: isn't it better to listen to the event wrapper (one of) being published, and do a switch here? Benefits are only one subscription, and making sure we handle all events.

File: Sources/StreamVideo/Call.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/Call.swift:932]: I don't want to make any other changes in this PR to avoid breaking it. In another PR is alright also because we are not targeting develop.

File: Sources/StreamVideo/Controllers/CallController.swift (hunk 33 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Controllers/CallController.swift:131]: We will as I mentioned earlier.

File: Sources/StreamVideo/CallKit/CallKitAdapter.swift (hunk 31 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/CallKit/CallKitAdapter.swift]: The change isn't showing. What was the issue?

**Review by martinmitrevski — COMMENTED** [review]
Looks good 👍  Let's discuss through the comments and I believe we can merge it tomorrow

File: Sources/StreamVideo/Utils/StateMachine/Publisher+NextValue.swift [hunk 35]
```diff
@@ -5,21 +5,45 @@
 import Combine
 import Foundation
 
-import Combine
-
 extension Publisher {
+
     /// Retrieves the next value from the publisher after optionally skipping the initial values.
     ///
     /// - Parameter dropFirst: The number of initial values to skip. Defaults to 0.
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/Utils/StateMachine/Publisher+NextValue.swift:12]: That will be done in another PR not here

File: Sources/StreamVideo/WebRTC/v2/CallAuthenticator.swift [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+protocol CallAuthenticating {
+
+    func authenticate(
+        create: Bool,
+        ring: Bool,
+        migratingFrom: String?
+    ) async throws -> JoinCallResponse
+}
+
+final class CallAuthenticator: CallAuthenticating {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/CallAuthenticator.swift:16]: That was mainly when we had both the WebRTCClient and the WebRTCCoordinator at the same time. We can probably remove it. A decision for when we reach a better state for the reconneciton.

File: DemoApp/Sources/Components/AppEnvironment.swift [hunk 37]
```diff
@@ -113,7 +113,7 @@ extension AppEnvironment {
         case .test:
             return .demo
         case .debug:
-            return .pronto
+            return .prontoStaging
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply DemoApp/Sources/Components/AppEnvironment.swift:116]: It's still useful for testing. As we are not merging into develop i suggest keeping it.

File: Sources/StreamVideo/WebRTC/v2/PeerConnection/Adapters/ICEAdapter.swift [hunk 38]
```diff
@@ -0,0 +1,278 @@
...
+            return
+        }
+        trickleTask(for: candidate)
+            .store(in: disposableBag)
+    }
+
+    private func trickleTask(
+        for candidate: RTCIceCandidate
+    ) -> Task<Void, Never> {
+        Task {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/PeerConnection/Adapters/ICEAdapter.swift:48]: I don't see any issues. The task will execute up to the point where we go to the SFUAdapter. At this point the actor will pick up something else to work on until the SFUAdapter returns. When we get the result from the SFUAdapter the actor will atomically update the trickedCandidates by only appending (not removing that's an important detail). Meaning that even in concurrent tasks trying to modify the array things will be as expected (indexes only increasing and we are relying on appending rather than pure indexes, so count doesn't matter).

File: Sources/StreamVideo/WebRTC/v2/SFU/SFUAdapter.swift [hunk 39]
```diff
@@ -0,0 +1,821 @@
...
+        var url: URL
+        /// The event notification center for handling WebSocket events.
+        var eventNotificationCenter: EventNotificationCenter
+        /// The session configuration for the WebSocket connection. Defaults to not waiting for connectivity.
+        var sessionConfiguration: URLSessionConfiguration = .default.toggleWaitsForConnectivity(false)
+        /// The decoder for WebRTC events. Defaults to a standard WebRTCEventDecoder.
+        var eventDecoder: WebRTCEventDecoder = WebRTCEventDecoder()
+    }
+
+    private let service: Stream_Video_Sfu_Signal_SignalServer
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/SFU/SFUAdapter.swift:40]: I prefer the word service of server and I can compromise with `signalService`.

File: Sources/StreamVideo/WebRTC/v2/WebRTCCoordinator.swift [hunk 40]
```diff
@@ -0,0 +1,330 @@
...
+    private func didTransition(
+        to stageId: StateMachine.Stage.ID
+    ) {
+        switch stageId {
+        default:
+            break
+        }
+    }
+
+    private func observeForceReconnectionRequests() {
```
[the remark this answers is not in the document]
Comment (ipavlidakis) [reply Sources/StreamVideo/WebRTC/v2/WebRTCCoordinator.swift:274]: We are going to remove this before we move to develop (or keep the implementation but remove the observation).
